### PR TITLE
[SES-PHASE-1-2] - PLU-744: add migration and email suppression model

### PR DESCRIPTION
**Pull request by m0nggh**
### TL;DR

Introduces an `email_suppression` table and `EmailSuppressionEntry` model to track emails that should not receive outbound mail due to SES permanent bounces or complaints.

### What changed?

- Added a new `email_suppression` database table via migration, storing the email address, suppression reason (`BOUNCE` or `COMPLAINT`), optional reason detail, SES message ID, whitelist count, and soft-delete/whitelist timestamps. A DB-level `CHECK` constraint enforces that only `BOUNCE` and `COMPLAINT` reasons are stored.
- Added the `EmailSuppressionEntry` Objection.js model with three static methods:
    - `upsertSuppression` — inserts or updates a suppression record from an SES event. On conflict, it overwrites the reason fields, re-suppresses the email, and increments `whitelist_count` only if the email had previously been whitelisted.
    - `getSuppressedEmails` — filters a list of emails down to those currently suppressed (i.e. present in the table with `last_whitelisted_at IS NULL`).
    - `whitelistEmails` — sets `last_whitelisted_at` on currently suppressed emails, effectively un-suppressing them without resetting the whitelist counter.

### How to test?

Nothing much to test because this just creates the new table and some functions.

- Run the integration tests in `email-suppression-entry.itest.ts`, which cover upsert idempotency, whitelist count incrementing, re-suppression after whitelisting, `getSuppressedEmails` filtering, and `whitelistEmails` behaviour.
- Run the unit tests in `email-suppression-entry.test.ts` for schema and empty-input edge cases.
- Apply the migration and verify the `email_suppression` table is created with the expected columns and `CHECK` constraint.

### Why make this change?

To prevent sending emails to addresses that have permanently bounced or filed spam complaints via SES. Continuing to send to such addresses harms sender reputation and violates email best practices. This model provides the persistence layer needed to honour those suppression signals and allow admins to override them when appropriate.

### CODE REVIEW

**Comment by m0nggh** [comment]
* **#1636** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1636?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1684** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1684?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1634** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1634?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1633** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1633?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1632** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1632?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1631** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1631?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1630** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1630?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1629** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1629?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1628** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1628?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1628?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1627** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1627?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feat/ses/trunk`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/backend/src/models/email-suppression-entry.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+      .insert({
+        email,
+        reason,
+        reasonDetail: reasonDetail ?? null,
+        sesMessageId: sesMessageId ?? null,
+        whitelistCount: 0,
+        lastWhitelistedAt: null,
+      })
+      .onConflict('email')
+      .merge({
```
Comment (pregnantboy) [inline packages/backend/src/models/email-suppression-entry.ts:131]: our queries are automatically appended with `DELETED_AT is null`. Check that this actually undeletes. the row, if not u might need to add .withSoftDeleted()

File: packages/backend/src/models/email-suppression-entry.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+ *  - 'virus'                    reported as virus
+ *  - 'other'                    unspecified
+ *
+ * reasonDetail is informational only — suppression checks only look at the
+ * row's existence + last_whitelisted_at, never at reason/reasonDetail.
+ */
+
+class EmailSuppressionEntry extends Base {
+  id!: string
+  email!: string
```
Comment (pregnantboy) [inline packages/backend/src/models/email-suppression-entry.ts:40]: is it possible to ensure the emails are lowercased before insertion? maybe add a beforeInsert/beforeUpdate hook to lower case email automatically in this model.

File: packages/backend/src/db/migrations/20260519100000_create_email_suppression.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    table.string('reason_detail', 50).nullable()
+    table.string('ses_message_id', 255).nullable()
+    table.integer('whitelist_count').notNullable().defaultTo(0)
+    table.timestamp('last_whitelisted_at', { useTz: true }).nullable()
+    table.timestamp('deleted_at', { useTz: true }).nullable()
+    table.timestamps(true, true) // created_at, updated_at with defaults
+
+    // Enforce the invariant at the DB layer — transient bounces and
+    // unknown event types must never end up in this table.
+    table.check("reason IN ('BOUNCE', 'COMPLAINT')")
```
Comment (pregnantboy) [inline packages/backend/src/db/migrations/20260519100000_create_email_suppression.ts:37]: what's this ah?

File: packages/backend/src/models/email-suppression-entry.ts (hunk 1 above)
Comment (pregnantboy) [reply packages/backend/src/models/email-suppression-entry.ts:131]: do we need to update updated_at as well?

**Review by pregnantboy — APPROVED** [review]
lost mostly ok, just some nits

File: packages/backend/src/db/migrations/20260519100000_create_email_suppression.ts (hunk 3 above)
Comment (m0nggh) [reply packages/backend/src/db/migrations/20260519100000_create_email_suppression.ts:37]: it adds a check on the reason column, just an additional layer on the db level ba (after i check with claude and verify on the structure for the `email_suppression` table

File: packages/backend/src/models/email-suppression-entry.ts (hunk 2 above)
Comment (m0nggh) [reply packages/backend/src/models/email-suppression-entry.ts:40]: good idea thks

File: packages/backend/src/models/email-suppression-entry.ts (hunk 1 above)
Comment (m0nggh) [reply packages/backend/src/models/email-suppression-entry.ts:131]: nothing in the codebase should actually set the deleted_at to non-null on this table. but ill add .withSoftDeleted() as a safeguard, thanks!
